### PR TITLE
[Fixes #9381] Hide users to anonoymous users and restric visibility for authenticated users

### LINKED ISSUE #9381: Hide users to anonoymous users and restric visibility for authenticated users
At the moment the user's page is visible to anyone, including anonymous users, which is quite odd and unexpected.
Also users visibility to authenticated users is not coherent between the APIs (legacy and new).

We want to align all the APIs and pages to the following rules:

 - anonymous users cannot access the `/people` and `/profile/<username>` pages, and cannot get list of registered users from the APIs
 - registered users can only see members of public groups and members of private groups the current user belongs to

The following exception applies to the visibility of users that are resource owners:

 - the current user (be it anonymous or registered) can see the list and the profiles of users that are owners of at least one resource that is visible to it

This exception is already implemented by the /owners API. I would keep this endpoint as the only implementation of this exception. The other endpoints (legacy and api/v2) will enforce the two rules above.

The affected APIs are:

  - [/api/v2/users](https://github.com/GeoNode/geonode/blob/e9364af46bf382dadb22ac8cbb5a278486151963/geonode/base/api/views.py#L115)
  - [/profiles](https://github.com/GeoNode/geonode/blob/e9364af46bf382dadb22ac8cbb5a278486151963/geonode/api/api.py#L455)
  - [/people/autocomplete](https://github.com/GeoNode/geonode/blob/e9364af46bf382dadb22ac8cbb5a278486151963/geonode/people/views.py#L143)

Access to  `/people` and `/profile/<username>`  pages:

 - apply the `login_required` decorator to the urls
 - do not render the Users menu item (this needs a change in the geonode client)

### PR DESCRIPTION
**Pull request by marthamareal**
<Include a few sentences describing the overall goals for this Pull Request>
References #9381 

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

For all pull requests:

- [x] Confirm you have read the [contribution guidelines](https://github.com/GeoNode/geonode/blob/master/CONTRIBUTING.md) 
- [x] You have sent a Contribution Licence Agreement (CLA) as necessary (not required for small changes, e.g., fixing typos in the documentation)
- [x] Make sure the first PR targets the master branch, eventual backports will be managed later. This can be ignored if the PR is fixing an issue that only happens in a specific branch, but not in newer ones.

The following are required only for core and extension modules (they are welcomed, but not required, for contrib modules):
- [x] There is a ticket in https://github.com/GeoNode/geonode/issues describing the issue/improvement/feature (a notable exemption is, changes not visible to end-users)
- [x] The issue connected to the PR must have Labels and Milestone assigned
- [x] PR for bug fixes and small new features are presented as a single commit
- [x] Commit message must be in the form "[Fixes #<issue_number>] Title of the Issue"
- [ ] New unit tests have been added covering the changes, unless there is an explanation on why the tests are not necessary/implemented
- [ ] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [ ] This PR passes the QA checks: flake8 geonode
- [ ] Commits changing the **settings**, **UI**, **existing user workflows**, or adding **new functionality**, need to include documentation updates
- [ ] Commits adding **new texts** do use gettext and have updated .po / .mo files (without location infos)

**Submitting the PR does not require you to check all items, but by the time it gets merged, they should be either satisfied or inapplicable.**

### CODE REVIEW

File: geonode/people/utils.py [hunk 1]
```diff
@@ -103,3 +106,19 @@ def format_address(street=None, zipcode=None, city=None, area=None, country=None
...
+
+def get_available_users(user):
+    if user.is_superuser:
+        return get_user_model().objects.exclude(Q(username='AnonymousUser') | Q(is_active=False))
+    # Only return user that are members of any group profile the current user is member of
+    member_ids = list(GroupMember.objects.filter(
+        group__in=GroupProfile.objects.filter(
+            Q(access='public') | Q(group__in=user.groups.all()))
+        ).select_related('user').values_list('user__id', flat=True))
+    if Group.objects.filter(name=groups_settings.REGISTERED_MEMBERS_GROUP_NAME).count():
```
Comment (mattiagiupponi) [inline geonode/people/utils.py:119]: Can we add a comment on why we always add the users that belong to the Registered member's groups? Just as an additional context

File: geonode/people/views.py [hunk 2]
```diff
@@ -145,6 +147,9 @@ class ProfileAutocomplete(autocomplete.Select2QuerySetView):
     def get_queryset(self):
         qs = get_user_model().objects.all().exclude(Q(username='AnonymousUser') | Q(is_active=False))
 
+        if self.request and self.request.user and not self.request.user.is_superuser:
```
Comment (mattiagiupponi) [inline geonode/people/views.py:150]: same as above is better to move the qs as an `else` condition to don't hit the db two times

**Review by mattiagiupponi — CHANGES_REQUESTED** [review]

File: geonode/people/utils.py [hunk 3]
```diff
@@ -103,3 +106,19 @@ def format_address(street=None, zipcode=None, city=None, area=None, country=None
         if country:
             address.append(country)
         return " ".join(address)
+
+
+def get_available_users(user):
+    if user.is_superuser:
+        return get_user_model().objects.exclude(Q(username='AnonymousUser') | Q(is_active=False))
```
[the remark this answers is not in the document]
Comment (giohappy) [reply geonode/people/utils.py]: IMHO no, If it wants there's the Django admin to see users in pending states.

File: geonode/people/utils.py (hunk 1 above)
Comment (marthamareal) [reply geonode/people/utils.py]: updated

Comment (marthamareal) [reply geonode/people/utils.py]: updated

**Review by mattiagiupponi — APPROVED** [review]

**Pull request merged**